json/rest/compose_data.go

```diff
@@ -0,0 +1,207 @@
+/*
+* Copyright(c) 2012-2021 Intel Corporation
+* SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+package rest
+
+import (
+	"fmt"
+
+	"example.com/json/ioctl"
+)
+
+/** packages of JSON marshalable structs with packed CAS info for easy use in RESTful API */
+
+type Cache_with_cores struct {
+	// default value of ptr is nil
+	Cache_info *ioctl.Kcas_cache_info  `json:"Cache"`
+	Cores_info []*ioctl.Kcas_core_info `json:"Cores"`
+}
+
+type Stats_pkg struct {
+	Cache_info *ioctl.Kcas_cache_info `json:"Cache info"`
+	Core_info  *ioctl.Kcas_core_info  `json:"Core id"`
+	Io_class   *ioctl.Kcas_io_class   `json:"IO class id"`
+	Stats      *ioctl.Kcas_get_stats  `json:"Statistics"`
+}
+
+type Cache_list_pkg []Cache_with_cores
+
+/** Retriving caches and their cores info from CAS */
+
+func List_caches() Cache_list_pkg {
+	var cache_list_pkg Cache_list_pkg
+
+	/** retrive file descriptor of exported object cas_ctrl */
+	fd, fd_err := ioctl.Get_cas_ctrl_fd()
+
+	if !check_err(fd_err, "cas_ctrl object access failed", log_console) {
+
+		/** retive list of caches id from ioclt */
+		cache_id_list, ioctl_err := get_cache_id_list(fd)
+
+		/**  create array of chaches with list of their cores */
+		cache_id_list_len := len(cache_id_list)
+		cache_list_pkg = make([]Cache_with_cores, cache_id_list_len)
+
+		if !check_err(ioctl_err, "could not retrive cache id list", log_console) {
+			for cache_idx, cache_id := range cache_id_list {
+
+				/** retrive cache info */
+				cache_info, ioctl_err := get_cache_info(fd, cache_id)
+
+				/** create array of cores for each cache */
+				cores_id_list_len := len(cache_info.Core_id)
+				cache_list_pkg[cache_idx].Cores_info = make([]*ioctl.Kcas_core_info, cores_id_list_len)
+
+				if !check_err(ioctl_err, "could not retrive cache info", log_console) {
+
+					/** fill array of caches & their cores with retrived cache info */
+					cache_list_pkg[cache_idx].Cache_info = &cache_info
+
+					for core_idx, core_id := range cache_info.Core_id {
+
+						/** retrive core info for table of cores assigned to cache */
+						core_info, ioctl_err := get_core_info(fd, cache_id, core_id)
+						check_err(ioctl_err, "could not retrive core info", log_console)
+
+						/** fill array of cores with retrived core info */
+						cache_list_pkg[cache_idx].Cores_info[core_idx] = &core_info
+					}
+				}
+			}
+		}
```

Comment:
In case of `get_cas_ctrl_fd` and `get_cache_id_list` errors early return would be better (it will be applied).
In case of `get_cache_info`, `get_core_info`I think better can be to store successfull values and errors and return them at the end. Idea of returning array []error and result with struct which we managed to retrive. It significantly decreases indentation.